sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/InsertAdaptiveSparkPlan.scala

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.adaptive
+
+import scala.collection.concurrent.TrieMap
+import scala.collection.mutable
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.expressions
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution._
+import org.apache.spark.sql.execution.command.ExecutedCommandExec
+import org.apache.spark.sql.execution.exchange.{Exchange, ShuffleExchangeExec}
+import org.apache.spark.sql.types.StructType
+
+/**
+ * This rule wraps the query plan with an [[AdaptiveSparkPlanExec]], which executes the query plan
+ * and re-optimize the plan during execution based on runtime data statistics.
+ */
+case class InsertAdaptiveSparkPlan(session: SparkSession) extends Rule[SparkPlan] {
+
+  private val conf = session.sessionState.conf
+
+  // Exchange-reuse is shared across the entire query, including sub-queries.
+  private val stageCache = new TrieMap[StructType, mutable.Buffer[(Exchange, QueryStageExec)]]()
+
+  override def apply(plan: SparkPlan): SparkPlan = plan match {
+    case _: ExecutedCommandExec => plan
+    case _ if conf.runtimeReoptimizationEnabled
+      && supportAdaptive(plan) =>
+      try {
+        // Plan sub-queries recursively and pass in the shared stage cache for exchange reuse. Fall
+        // back to non-adaptive mode if adaptive execution is supported in any of the sub-queries.
+        val subqueryMap = planSubqueries(plan)
+        // Run preparation rules.
+        val preparations = AdaptiveSparkPlanExec.createQueryStagePreparationRules(
+          session.sessionState.conf, subqueryMap)
+        val newPlan = AdaptiveSparkPlanExec.applyPhysicalRules(plan, preparations)
```

Comment:
Why do we need to do this? This already seems to be done in `AdaptiveSparkPlanExec` when we submit the stage.

Reply:
> This already seems to be done in AdaptiveSparkPlanExec when we submit the stage.
No.

The physical transformations we used in `QueryExecution.preparations` have now been split into two groups in adaptive execution here (also noted in the code comment):
1. Rules that add or remove exchanges.
2. Rules that are independent within each exchange, or say, stage.

`InsertAdaptiveSparkPlan` is now the first in `QueryExecution.preparations`, which means neither of these two groups has been applied yet. It is this way so that we do not need to manipulate (modify) the rule application order in `QueryExecution.preparations` for AQE.

Reply:
To me it makes more sense to this in the `AdaptiveSparkPlanExec`. The `AdaptiveSparkPlanExec` is now expecting a plan that can only be produced by this rule, and not any physical plan.

Reply:
In one of the commits I had tried refactoring this into `AdaptiveSparkPlanExec`, but later found out that this would cause a problem in serializing `initialPlan` in `AdaptiveSparkPlanExec`, for the `initialPlan` before applying the sub-query planning rule contains instances of `expression.ScalarSubquery`.